src/main/haskell/kore/src/Kore/Step/Function/Registry.hs

```diff
@@ -30,40 +29,34 @@ import Kore.AST.PureML
 import Kore.AST.Sentence
        ( AsSentence (..), SentenceAxiom (..) )
 import Kore.IndexedModule.IndexedModule
-       ( IndexedModule (..), KoreIndexedModule )
+       ( IndexedModule (..), KoreIndexedModule, indexedModulesInScope )
 import Kore.Step.AxiomPatterns
-       ( AxiomPattern (..), QualifiedAxiomPattern (..),
-       koreSentenceToAxiomPattern )
 import Kore.Step.Function.Data
        ( ApplicationFunctionEvaluator (..) )
 import Kore.Step.Function.UserDefined
        ( axiomFunctionEvaluator )
 import Kore.Step.StepperAttributes
        ( StepperAttributes )
 
-{-|Given a 'MetaOrObject' @level@ and a 'KoreIndexedModule',
-creates a registry mapping function symbol identifiers to their
-corresponding 'AxiomPattern's.
--}
+{- | Create a mapping from symbol identifiers to their defining axioms.
+
+ -}
 extractFunctionAxioms
     :: MetaOrObject level
     => level
     -> KoreIndexedModule StepperAttributes
     -> Map.Map (Id level) [AxiomPattern level]
-extractFunctionAxioms level indexedModule =
-    Map.fromList (map extractPrefix groupedAxiomPatterns)
+extractFunctionAxioms level =
+    \imod -> execState (extractFunctionAxioms0 imod) Map.empty
   where
-    extractPrefix []                  = error "unexpected case"
-    extractPrefix ((a, b) : reminder) = (a, b : map snd reminder)
-    groupedAxiomPatterns =
-        groupBy
-            ((==) `on` fst)
-            (sortBy
-                (compare `on` fst)
-                (mapMaybe
-                    (axiomToIdAxiomPatternPair level)
-                    (snd <$> indexedModuleAxioms indexedModule))
-            )
+    extractFunctionAxioms0 imod =
+        mapM_ (mapM_ extractAxiomPattern) axiomsInScope
+      where
+        axiomsInScope = indexedModuleAxioms <$> indexedModulesInScope imod
+    insertAxiom (name, patt) =
+        Monad.State.modify' $ Map.alter (Just . (patt :) . fromMaybe []) name
+    extractAxiomPattern (_, axiomSentence) =
+        mapM_ insertAxiom (axiomToIdAxiomPatternPair level axiomSentence)
```

Comment:
While I can agree that the refactoring above makes it more compact (and perhaps more efficient, due to using the `State` monad), I'm not sure about it being more readable.  
Could you please at least add signatures for the functions (esp. for the ones involving `mapM_`) so one would follow more easily what is `mapM_` mapping on?  I found the one for pairs a little strange, but I guess I could get used to it :-)